src/backend/distributed/transaction/transaction_management.c

```diff
@@ -731,36 +752,22 @@ AdjustMaxPreparedTransactions(void)
 static void
 PushSubXact(SubTransactionId subId)
 {
-	/*
-	 * We need to allocate these in TopTransactionContext instead of current
-	 * subxact's memory context. This is because AtSubCommit_Memory won't
-	 * delete the subxact's memory context unless it is empty, and this
-	 * can cause in memory leaks. For emptiness it just checks if the memory
-	 * has been reset, and we cannot reset the subxact context since other
-	 * data can be in the context that are needed by upper commits.
-	 *
-	 * See https://github.com/citusdata/citus/issues/3999
-	 */
-	MemoryContext old_context = MemoryContextSwitchTo(TopTransactionContext);
-
 	/* save provided subId as well as propagated SET LOCAL stmts */
 	SubXactContext *state = palloc(sizeof(SubXactContext));
 	state->subId = subId;
 	state->setLocalCmds = activeSetStmts;
 
 	/* append to list and reset active set stmts for upcoming sub-xact */
-	activeSubXactContexts = lcons(state, activeSubXactContexts);
+	activeSubXactContexts = lappend(activeSubXactContexts, state);
 	activeSetStmts = makeStringInfo();
-
-	MemoryContextSwitchTo(old_context);
 }
 
 
 /* PopSubXact pops subId from the stack of active sub-transactions. */
 static void
 PopSubXact(SubTransactionId subId)
 {
-	SubXactContext *state = linitial(activeSubXactContexts);
+	SubXactContext *state = llast(activeSubXactContexts);
```

Comment:
Why do we change these operations in this PR? Is that to avoid `reversedSubXactStates` ? I'm asking because some of the comments now seems to not fit as good as before.

In any case, I'd probably make these change in a follow-up PR because felt like we might have missed something